src/pint/gridutils.py

```diff
@@ -1,16 +1,400 @@
+    |                 |        | ``cpu_count`` workers  |
+    +-----------------+--------+------------------------+
+    | None	      | >1     | creates default        |
+    |                 |        | executor with desired  |
+    |                 |        | number of workers      |
+    +-----------------+--------+------------------------+
+    """
+
+    if isinstance(executor, concurrent.futures.Executor):
+        # the executor has already been created
+        executor = executor
+    elif executor is None and (ncpu is None or ncpu > 1):
+        # make the default type of Executor
+        if ncpu is None:
+            ncpu = multiprocessing.cpu_count()
+        executor = concurrent.futures.ProcessPoolExecutor(max_workers=ncpu)
+
+    # Save the current model so we can tweak it for gridding, then restore it at the end
+    savemod = ftr.model
+    gridmod = copy.deepcopy(ftr.model)
+    ftr.model = gridmod
+
+    # Freeze the  params we are going to grid over
+    for parname in parnames:
+        getattr(ftr.model, parname).frozen = True
+
+    # All other unfrozen parameters will be fitted for at each grid point
+    out = np.meshgrid(*parvalues)
+    chi2 = np.zeros(out[0].shape)
+    # at this point, if the executor is None then run single-processor version
+    if executor is not None:
+        with executor as e:
+            if printprogress and tqdm is not None:
+                result = list(
+                    tqdm(
+                        e.map(
+                            doonefit,
+                            (ftr,) * len(out[0].flatten()),
+                            (parnames,) * len(out[0].flatten()),
+                            list(zip(*[x.flatten() for x in out])),
+                            chunksize=chunksize,
+                        ),
+                        total=len(out[0].flatten()),
+                        ascii=True,
+                    )
+                )
+            else:
+                result = e.map(
+                    doonefit,
+                    (ftr,) * len(out[0].flatten()),
+                    (parnames,) * len(out[0].flatten()),
+                    list(zip(*[x.flatten() for x in out])),
+                    chunksize=chunksize,
+                )
+        it = np.ndindex(chi2.shape)
+        for i, r in zip(it, result):
+            chi2[i] = r
+    else:
+        it = np.ndindex(out[0].shape)
+        if printprogress and tqdm is not None:
+            it = tqdm(it, total=len(out[0].flatten()), ascii=True)
+        for i in it:
+            for parnum, parname in enumerate(parnames):
+                getattr(ftr.model, parname).quantity = out[parnum][i]
+            ftr.fit_toas()
+            chi2[i] = ftr.resids.chi2
+
+    # Restore saved model
+    ftr.model = savemod
+    return chi2
+
+
+def grid_chisq_derived(
+    ftr,
+    parnames,
+    parfuncs,
+    gridvalues,
+    executor=None,
+    ncpu=None,
+    chunksize=1,
+    printprogress=True,
+):
```

Comment:
This seems to involve a lot of code duplication - can the simpler interface be implemented on top of this one?